sdk/eventhub/Azure.Messaging.EventHubs.Shared/src/BlobCheckpointStore/BlobsCheckpointStore.cs

```diff
@@ -322,13 +339,82 @@ await foreach (BlobItem blob in ContainerClient.GetBlobsAsync(traits: BlobTraits
+                {
+                    // Skip new checkpoints and empty blobs
+                    if (blob.Properties.ContentLength == 0)
+                    {
+                        continue;
+                    }
+
+                    var partitionId = blob.Name.Substring(legacyPrefix.Length);
+
+                    // Check whether there is already a checkpoint for this partition id
+                    if (existingCheckpoints.Any(existingCheckpoint => string.Equals(existingCheckpoint.PartitionId, partitionId, StringComparison.Ordinal)))
+                    {
+                        continue;
+                    }
+
+                    var startingPosition = default(EventPosition?);
+
+                    BlobBaseClient blobClient = ContainerClient.GetBlobClient(blob.Name);
+                    using var memoryStream = new MemoryStream();
+                    await blobClient.DownloadToAsync(memoryStream, listCheckpointsToken).ConfigureAwait(false);
+
+                    TryReadLegacyCheckpoint(
+                        memoryStream.GetBuffer().AsSpan(0, (int)memoryStream.Length),
+                        out long? offset,
+                        out long? sequenceNumber);
+
+                    if (offset.HasValue)
+                    {
+                        startingPosition = EventPosition.FromOffset(offset.Value, false);
+                    }
+                    else if (sequenceNumber.HasValue)
+                    {
+                        startingPosition = EventPosition.FromSequenceNumber(sequenceNumber.Value, false);
+                    }
+
+                    if (startingPosition.HasValue)
+                    {
+                        checkpoints.Add(new BlobStorageCheckpoint
+                        {
+                            FullyQualifiedNamespace = fullyQualifiedNamespace,
+                            EventHubName = eventHubName,
+                            ConsumerGroup = consumerGroup,
+                            PartitionId = partitionId,
+                            StartingPosition = startingPosition.Value,
+                            Offset = offset,
+                            SequenceNumber = sequenceNumber
+                        });
+                    }
+                    else
+                    {
+                        InvalidCheckpointFound(partitionId, fullyQualifiedNamespace, eventHubName, consumerGroup);
+                    }
+                }
+
+                return checkpoints;
+            };
+
+            List<EventProcessorCheckpoint> checkpoints = null;
             try
             {
-                return await ApplyRetryPolicy(listCheckpointsAsync, cancellationToken).ConfigureAwait(false);
+
+                checkpoints = await ApplyRetryPolicy(listCheckpointsAsync, cancellationToken).ConfigureAwait(false);
+                if (ReadLegacyCheckpoints)
```

Comment:
Wouldn't this continue to make the extra query after all T1 checkpoints were converted?    I don't have a good answer on how to mitigate that, outside of passing in the set of partitions and looking to see that there was a T2 checkpoint found for each....

Reply:
Yes, that's why I was looking into GetCheckpointAsync implementation that takes partition id. That way we can avoid reading T1 when T2 already exists.

Reply:
Fair enough.   I've no objections if you leave as-is for now, and I'll give it a closer look when I look at adding per-partition calls.

Reply:
Filed https://github.com/Azure/azure-sdk-for-net/issues/17386